### ISSUE TITLE
Invalid type hint not detected for decorators on classes

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#8945]
**Describe the bug**
Pyright usually provides errors for invalid type hints, such as a function being used as a type hint, but if a class is wrapped by a decorator which returns a function, no error is thrown. Most of the time, it doesn't directly matter if the type hint is invalid, but if it's optional, like `foo: func | None`, then a `TypeError` will be thrown at runtime, despite no error being thrown by Pyright.

**Code or Screenshots**
[Playground to reproduce](https://pyright-play.net/?code=CYUwZgBAxgrgzgFwPYFsAUUA2cCUAuAKAmIlEigENNM0AqCgJwHM4AaCW2gawHdGX8REsIYgEMBgDto2OvzYdufZriHFR4qdCqYCBAAKxEqAlgpw4EAMIALAJaZghYQDo3eqPccBGPNHjIKBAAPhAAckiSIMQAxBAAogwMSAykYiBQCCDAEABGGRTw0Qg20faSCBB2lhQQYDCSmXaREGgAKgCeAA4gicmpFJUMDQh2KCA4pl7AAEx%2Btg45oRFRseFIECBJKWlZmdnsuTCVJdVVNZkwOh0QtfWNoy3t3b3bA0MjYxNAA)

```python
def custom(cls):
    def call(*args, **kwargs):
        return cls(*args, **kwargs)
    return call

@custom
class Child:
    ...

child1: custom | None  # Error detected because the hint is a function (TypeError at runtime)
child2: Child | None  # No error detected, but this is actually a function (TypeError at runtime)
```

**VS Code extension or command-line**
Found in Pyright 1.1.376 on commandline & LSP, reproduced on 1.1.397.

**Comment by erictraut** [comment]
This is by design. If you apply an unannotated decorator to a class, pyright assumes that it has no effect. This is typically the case, so it's a good assumption to make for untyped code bases. In the past, I've tried to use inferred return types for unannotated class decorators, but this leads to worse behaviors — especially for users who are not interested in type checking and are using pyright as a language server.

If you add type annotations to your `custom` decorator, you'll see the behavior you're looking for.
